specification/flag-evaluation/hooks.md

```diff
@@ -2,4 +2,75 @@
 
 ## Overview
 
-// TODO: this is a stub.
+Hooks are a mechanism whereby application developers can add arbitrary behavior
+to flag evaluation. They operate similarly to middleware in many web frameworks.
+
+### Definitions
+
+**Stage**: An explicit portion of the flag evaluation lifecycle. e.g. `before` being "before the evaluation is run.
+**Invocation**: A single call to evaluate a flag. `client.getBooleanValue(..)` is an invocation.
+**API**: The global API singleton.
+
+### Hook context
+
+Hook context exists to provide hooks with information about the invocation.
+
+> Hook context **MUST** provide: the flag key, flag type, evaluation context, default value and a list of  executed hooks by stage.
+
+> Hook context **SHOULD** provide: provider, client
+
+> flag key, flag type, default value properties **MUST** be immutable.
+
+> The evaluation context **MUST** be mutable only within the `before` hook.
+
+### State
```

Comment:
What is the exact use case of `State`? One question I have is does it risk causing flags to evaluate differently on different instances of a service?

Reply:
I'm primarily interested in, on a per-invocation basis, specifying information to a hook. Imagine I have a hook that emits metrics. I want that in almost all cases. I'd like the ability to say "For this particular request.. don't emit metrics".   I'm thinking we use `state` for that.

Reply:
That makes sense, I'm wondering if there's a way we can clarify or scope this expected usage down. We're quickly layering on top of the evaluation process, hooks, then state without an initial release.

Reply:
I'm OK with the concept, and I think I see the value from other discussions with @justinabrahms . It might be that we need an example use-case or other form of explanation (and perhaps to change the name to something that more clearly defines the purpose; "state" may be a bit vague?).

Once this is formatted to a spec document, that supplementary info (and even a code sample, perhaps) can be added in a non-normative section.

Reply:
@beeme1mr I have a very concrete use-case for it. I'd _prefer_ to keep it in this spec, as when it merges I'm going to update the java sdk code to do the thing I'm saying and then I can use the code internally. That said, if you feel strongly, I'll move it out.

The explicit use-case here is I have a metrics service. Sometimes, we want to control precisely when metrics are emitted (due to some pre-fetching logic we have), but usually not. I want to register a "please emit the relevant metrics" hook which triggers by default.

Sometimes, when I want to control when those metrics emit, I want to not have that piece of code run. To do that, I have to provide some state to the function. The options for state that I've considered:

1. We could use `evaluatedHooks`. I could register a `DoNotEmitHooks` hook, which is a sentinel value that `EmitHooks` looks for. If it sees it in the evaluation order before it.. it won't emit hooks. (eew).
2. We could use `EvaluationContext` which is the map of data we send up for evaluation. It could have users put data in there to communicate to the hooks, but isn't really used for targeting. (eew)
3. We could pass some explicit hook-specific state. That's what I'm proposing here.
4. We could call methods on the `EmitMetricsHook` itself like `dontEmitMetricsOnNextCall`, but I worry about the thread safety of this approach.
5. Maybe a secret 5th option that you can think of?

@moredip I hear you about rapidly iterating. This came up when we were trying to port one of the hooks we care about into the janky java sdk I wrote. We didn't actually have the relevant primitives to accomplish the goal, hence the proposal.

Reply:
I found the work "state" a bit misleading. I assumed it was a place for hooks to maintain state that could then be picked up by the hook in a later evaluation stage. I think what's being describe here is more like per-invocation configuration for a hook. Assuming I'm reading that correctly, then maybe "Hook Hints" would be a clearer name than state?

Reply:
@moredip  Agreed, that's really my primary concern with this requirement as well.

Reply:
Moved to `HookHints`. Much better name. :) Can we resolve this thread?

Reply:
👍 on resolving